File: fn/src/test/java/org/sqlbuilder/fn/TestCollector.java
package org.sqlbuilder.fn;

import org.junit.Test;
import org.sqlbuilder.common.SetCollector;

import java.util.Comparator;
import java.util.Set;

public class TestCollector
{
	private static final Set<String> SET = Set.of("one", "two", "three");

	private static final Comparator<String> COMPARATOR = Comparator.naturalOrder();

	private static final SetCollector<String> C = new SetCollector<>(COMPARATOR);

	private static final SetCollector<String> D = new SetCollector<>(COMPARATOR);

	@Test
	public void test()
	{
		for (var e : SET)
		{
			C.add(e);
		}

		System.err.println("[BEFORE]" + C.status());
		try (var ignored = C.open())
		{
			System.err.println("[ACTIVE]" + C.status());
			System.out.println(C.apply("one"));
			System.out.println(C.apply("two"));
			System.out.println(C.apply("three"));
			System.out.println(C.apply("four"));
		}
		System.err.println("[AFTER]" + C.status());
		try { System.out.println(C.apply("one")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println(C.apply("two")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println(C.apply("three")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println(C.apply("four")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
	}

	@Test(expected=IllegalStateException.class)
	public void testFail()
	{
		for (var e : SET)
		{
			C.add(e);
		}

		System.err.println("[BEFORE]" + C.status());
		try (var ignored = C.open())
		{
			System.err.println("[ACTIVE]" + C.status());
			System.out.println(C.apply("one"));
			System.out.println(C.apply("two"));
			System.out.println(C.apply("three"));
			System.out.println(C.apply("four"));
		}
		System.err.println("[AFTER]" + C.status());
		System.out.println(C.apply("one"));
	}

	@Test
	public void test2()
	{
		for (var e : SET)
		{
			C.add(e);
			D.add(e);
		}

		System.err.println("[BEFORE]" + C.status());
		try (var ignored = C.open(); var ignored2 = D.open())
		{
			System.err.println("[ACTIVE]" + C.status() + D.status());
			System.out.println("c " + C.apply("one"));
			System.out.println("c " + C.apply("two"));
			System.out.println("c " + C.apply("three"));
			System.out.println("c " + C.apply("four"));
			System.out.println("d " + D.apply("one"));
			System.out.println("d " + D.apply("two"));
			System.out.println("d " + D.apply("three"));
			System.out.println("d " + D.apply("four"));
		}
		System.err.println("[AFTER]" + C.status());
		try { System.out.println("c " + C.apply("one")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("c " + C.apply("two")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("c " + C.apply("three")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("c " + C.apply("four")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("d " + D.apply("one")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("d " + D.apply("two")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("d " + D.apply("three")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
		try { System.out.println("d " + D.apply("four")); } catch (IllegalStateException ise) {System.out.println(ise.getMessage());}
	}

	@Test(expected=IllegalStateException.class)
	public void test2Fail()
	{
		for (var e : SET)
		{
			C.add(e);
			D.add(e);
		}

		System.err.println("[BEFORE]" + C.status());
		try (var ignored = C.open(); var ignored2 = D.open())
		{
			System.err.println("[ACTIVE]" + C.status() + D.status());
			System.out.println("c " + C.apply("one"));
			System.out.println("c " + C.apply("two"));
			System.out.println("c " + C.apply("three"));
			System.out.println("c " + C.apply("four"));
			System.out.println("d " + D.apply("one"));
			System.out.println("d " + D.apply("two"));
			System.out.println("d " + D.apply("three"));
			System.out.println("d " + D.apply("four"));
		}
		System.err.println("[AFTER]" + C.status());
		System.out.println("c " + C.apply("one"));
	}
}

File: fn/src/test/java/org/sqlbuilder/fn/TestCollector.kt
package org.sqlbuilder.fn

import org.junit.Test
import org.sqlbuilder.common.SetCollector

class TestCollector {

    @Test
    fun test() {
        for (e in SET) {
            C.add(e)
        }

        System.err.println("[BEFORE]" + C.status())
        C.open().use {
            System.err.println("[ACTIVE]" + C.status())
            println(C.invoke("one"))
            println(C.invoke("two"))
            println(C.invoke("three"))
            println(C.invoke("four"))
        }
        System.err.println("[AFTER]" + C.status())
        try {
            println(C.invoke("one"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println(C.invoke("two"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println(C.invoke("three"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println(C.invoke("four"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
    }

    @Test(expected = IllegalStateException::class)
    fun testFail() {
        for (e in SET) {
            C.add(e)
        }

        System.err.println("[BEFORE]" + C.status())
        C.open().use {
            System.err.println("[ACTIVE]" + C.status())
            println(C.invoke("one"))
            println(C.invoke("two"))
            println(C.invoke("three"))
            println(C.invoke("four"))
        }
        System.err.println("[AFTER]" + C.status())
        println(C.invoke("one"))
    }

    @Test
    fun test2() {
        for (e in SET) {
            C.add(e)
            D.add(e)
        }

        System.err.println("[BEFORE]" + C.status())
        C.open().use {
            D.open().use {
                System.err.println("[ACTIVE]" + C.status() + D.status())
                println("c " + C.invoke("one"))
                println("c " + C.invoke("two"))
                println("c " + C.invoke("three"))
                println("c " + C.invoke("four"))
                println("d " + D.invoke("one"))
                println("d " + D.invoke("two"))
                println("d " + D.invoke("three"))
                println("d " + D.invoke("four"))
            }
        }
        System.err.println("[AFTER]" + C.status())
        try {
            println("c " + C.invoke("one"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("c " + C.invoke("two"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("c " + C.invoke("three"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("c " + C.invoke("four"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("d " + D.invoke("one"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("d " + D.invoke("two"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("d " + D.invoke("three"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
        try {
            println("d " + D.invoke("four"))
        } catch (ise: IllegalStateException) {
            println(ise.message)
        }
    }

    @Test(expected = IllegalStateException::class)
    fun test2Fail() {
        for (e in SET) {
            C.add(e)
            D.add(e)
        }

        System.err.println("[BEFORE]" + C.status())
        C.open().use {
            D.open().use {
                System.err.println("[ACTIVE]" + C.status() + D.status())
                println("c " + C.invoke("one"))
                println("c " + C.invoke("two"))
                println("c " + C.invoke("three"))
                println("c " + C.invoke("four"))
                println("d " + D.invoke("one"))
                println("d " + D.invoke("two"))
                println("d " + D.invoke("three"))
                println("d " + D.invoke("four"))
            }
        }
        System.err.println("[AFTER]" + C.status())
        println("c " + C.invoke("one"))
    }

    companion object {

        private val SET = mutableSetOf<String>("one", "two", "three", "four")

        private val COMPARATOR: Comparator<String> = Comparator.naturalOrder()

        private val C = SetCollector<String>(COMPARATOR)

        private val D = SetCollector<String>(COMPARATOR)
    }
}
